pep-0646.rst

```diff
@@ -743,11 +753,199 @@ is available in `cpython/23527`_. A preliminary version of the version
+
+::
+
+    def f(x: int): ...
+   
+Note that we have symbols for both the value of the thing (``x``) and the type of
+the thing (``int``). Why can't we do the same with axes? For example, with an imaginary
+syntax, we could write:
+
+::
+
+    def f(array: Array[TimeValue: TimeType]): ...
+    
+This would allow us to access the axis size (say, 32) through the symbol ``TimeValue``
+*and* the type through the symbol ``TypeType``.
+
+This might even be possible using existing syntax, through a second level of parameterisation:
+
+::
+
+   def f(array: array[TimeValue[TimeType]]): ..
+   
+However, we leave exploration of this approach to the future.
+
+Appendix B: Shaped Types vs Named Axes
+======================================
+
+An issue related to those addressed by this PEP concerns
+axis *selection*. For example, if we have an image stored in an array of shape 64×64x3,
+we might wish to convert to black-and-white by computing the mean over the third
+axis, ``mean(image, axis=2)``. Unfortunately, the simple typo ``axis=1`` is
+difficult to spot and will produce a result that means something completely different
+(all while likely allowing the program to keep on running, resulting in a bug
+that is serious but silent).
+
+In response, some libraries have implemented so-called 'named tensors' (in this context,
+'tensor' is synonymous with 'array'), in which axes are selected not by index but by
+label - e.g. ``mean(image, axis='channels')``.
+
+A question we are often asked about this PEP is: why not just use named tensors?
+The answer is that we consider the named tensors approach insufficient, for two main reasons:
+
+* **Static checking** of shape correctness is not possible. As mentioned in `Motivation`_,
+  this is a highly desireable feature in machine learning code where iteration times
+  are slow by default.
+* **Interface documentation** is still not possible with this approach. If a function should
+  *only* be willing to take array arguments that have image-like shapes, this cannot be stipulated
+  with named tensors.
+
+Additionally, there's the issue of **poor uptake**. At the time of writing, named tensors
+have only been implemented in a small number of numerical computing libraries. Possible explanations for this
+include difficulty of implementation (the whole API must be modified to allow selection by axis name
+instead of index), and lack of usefulness due to the fact that axis ordering conventions are often
+strong enough that axis names provide little benefit (e.g. when working with images, 3D tensors are
+basically *always* height × width × channels). However, ultimately we are still uncertain
+why this is the case.
+
+Can the named tensors approach be combined with the approach we advocate for in
+this PEP? We're not sure. One area of overlap is that in some contexts, we could do, say,
+``image: Array[Height, Width, Channels]; mean(image, axis=Image.axes.index(Channels))``. Ideally,
```

Comment:
IMO this really needs separate lines, or perhaps:
```suggestion
``Image = Array[Height, Width, Channels]; image: Image; mean(image, axis=Image.axes.index(Channels))``. Ideally,
```

Right now `Image` is not defined.

Reply:
Oh, whoops, I meant to type lowercase `image` - `axis=image.axes.index(Channels)`. I was imagining each instance of `Array` would have an internal map `axes`, mapping from labels like `Channels` to axis indexes. Does that make sense?

Reply:
Ah, OK. You were imaging `axes` would part of the array objects. I was imagining it would just be part of the types.

There are certainly use-cases for both. In my opinion the advantage of keeping the information on the type objects is that you don't need to switch user-code to use new array types, hence lowering barriers to entry.

If you do want to pull it off the array object you should show how it gets there, e.g., `image = LabeledArray(data, axes=[Height, Width, Channels])`.

Reply:
Ohh, gotcha. Neat! I've made the change you suggested.